### PR TITLE
Script Editor UX for Level Builder

### PR DESCRIPTION
**Pull request by joshlory**
Replace the DSL text area with a UI for editing Stage and Levels in a Course Script.

screen shot 2016-11-09 at 3 41 47 pm

### CODE REVIEW

File: apps/src/sites/studio/pages/components/FlexGroup.jsx [hunk 1]
```diff
@@ -0,0 +1,113 @@
+import React from 'react';
+import { connect } from 'react-redux';
+import _ from 'lodash';
+import color from '../../../../util/color';
```
Comment (Bjvanminnen) [inline apps/src/sites/studio/pages/components/FlexGroup.jsx]: I wonder if we should start using `@cdo/apps/utils/color` instead of these long paths of `../` that are hard to follow. I think @pcardune had suggested that a good heuristic might be to provide relative paths for things below you (i.e. if your path starts with `./` but use `@cdo/apps` otherwise. Should we formalize that? (Not sure if there are downsides to using @cdo/apps in mroe places.

File: apps/src/sites/studio/pages/components/FlexGroup.jsx [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+const FlexGroup = React.createClass({
+  propTypes: {
+    addGroup: React.PropTypes.func.isRequired,
+    addStage: React.PropTypes.func.isRequired,
+    stages: React.PropTypes.array.isRequired
+  },
+
+  handleAddGroup() {
+    this.props.addGroup();
+  },
```
Comment (Bjvanminnen) [inline apps/src/sites/studio/pages/components/FlexGroup.jsx]: I think you could just do `onMouseDown={this.props.addGroup}` instead of creating a wrapper method.

File: apps/src/sites/studio/pages/components/FlexGroup.jsx (hunk 1 above)
Comment (pcardune) [reply apps/src/sites/studio/pages/components/FlexGroup.jsx]: The main downside to using `@cdo/apps` in more places is that static analysis and refactoring tools are less likely to understand what they mean (as that would require them to understand webpack module aliases)

File: apps/src/sites/studio/pages/components/FlexGroup.jsx [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+                  afterStage++;
+                  return (
+                    <StageCard
+                      key={stage.id}
+                      stagesCount={this.props.stages.length}
+                      stage={stage}
+                    />
+                  );
+                })}
+                <button onMouseDown={this.handleAddStage.bind(null, afterStage - 1)} className="btn" style={styles.addStage} type="button">
```
Comment (Bjvanminnen) [inline apps/src/sites/studio/pages/components/FlexGroup.jsx]: styleguide prefers one prop per line when we have multiple props

Comment (Bjvanminnen) [reply apps/src/sites/studio/pages/components/FlexGroup.jsx]: What does "type=button" accomplish on a button?

File: apps/src/sites/studio/pages/components/FlexGroup.jsx [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+                  <i style={{marginRight: 7}} className="fa fa-plus-circle" />
```
Comment (Bjvanminnen) [inline apps/src/sites/studio/pages/components/FlexGroup.jsx]: Could use our FontAwesome component here and elsewhere

File: apps/src/sites/studio/pages/components/FlexGroup.jsx [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+      </div>
+    );
+  }
+});
+
+export default connect(state => ({
+  stages: state
+}), dispatch => ({
+  addGroup() {
+    dispatch({type: 'ADD_GROUP'});
```
Comment (Bjvanminnen) [inline apps/src/sites/studio/pages/components/FlexGroup.jsx]: The pattern we've been following elsewhere is that instead of dispatching a type with a string here, we have an action creator and a const in our redux file, i.e.

```
// editorRedux.js
const ADD_GROUP = 'editor/ADD_GROUP';
export function addGroup() {
  return { type: ADD_GROUP };
}

export function reducer(state=initialState, action) {
  if (action.type === ADD_GROUP) {
    return nextState;
  }
} 
```

**Comment by pcardune** [comment]
The `src/sites/studio/pages/components` directory should not exist. Any generic react components you create should live in `src/templates/`. Non-generic react components (or really any code that is super specific to script editor should live in `src/lib/script-editor`. (i know there isn't a lib directory in `src/` yet, but I will be adding one and moving a bunch of stuff there in the not-to-distant future.

File: apps/src/sites/studio/pages/components/LevelToken.jsx [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    this.props.toggleExpand(this.props.stagePosition, this.props.level.position);
+  },
+
+  handleRemove() {
+    this.props.removeLevel(this.props.stagePosition, this.props.level.position);
+  },
+
+  render() {
+    return (
+      <Motion
```
Comment (Bjvanminnen) [inline apps/src/sites/studio/pages/components/LevelToken.jsx]: There's enough complexity going on here that it might be worth a short comment.

File: apps/src/sites/studio/pages/components/LevelToken.jsx [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+    border: '1px solid #a00',
+    boxShadow: 'inset 0 1px 0 0 rgba(255, 255, 255, 0.6)',
+    padding: '7px 13px',
+    borderTopRightRadius: borderRadius,
+    borderBottomRightRadius: borderRadius,
+    cursor: 'pointer'
+  }
+};
+
+const LevelToken = React.createClass({
```
Comment (Bjvanminnen) [inline apps/src/sites/studio/pages/components/LevelToken.jsx]: What is a LevelToken? I'm not always great about this either, but I think would be good for us to have class level comments for most of our components describing what they are.

File: apps/src/sites/studio/pages/components/LevelToken.jsx [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+      >
+        {({y, scale, shadow}) =>
+          <div
+            style={Object.assign({}, styles.levelToken, {
+              transform: `translate3d(0, ${y}px, 0) scale(${scale})`,
+              boxShadow: `${color.shadow} 0 ${shadow}px ${shadow * 3}px`,
+              zIndex: this.props.drag ? 1000 : 500 - this.props.level.position
+            })}
+          >
+            <div style={styles.reorder} onMouseDown={this.props.handleDragStart.bind(null, this.props.level.position)}>
```
Comment (Bjvanminnen) [inline apps/src/sites/studio/pages/components/LevelToken.jsx]: This might be a place where it makes sense to have a class level method (similar to toggleExpand) such that we avoid rebinding on every render.

File: apps/src/sites/studio/pages/components/LevelTokenDetails.jsx [hunk 9]
```diff
@@ -0,0 +1,153 @@
+/* global levelKeyList */
+
+import React from 'react';
+import { connect } from 'react-redux';
+import VirtualizedSelect from 'react-virtualized-select';
+import 'react-virtualized/styles.css';
+import 'react-select/dist/react-select.css';
+import 'react-virtualized-select/styles.css';
```
Comment (Bjvanminnen) [inline apps/src/sites/studio/pages/components/LevelTokenDetails.jsx]: Is this the first place we're importing css? I'm a little surprised our webpack even knows how to do this.

File: apps/src/sites/studio/pages/components/editorRedux.js [hunk 10]
```diff
@@ -0,0 +1,127 @@
+/* global levelKeyList */
```
Comment (Bjvanminnen) [inline apps/src/sites/studio/pages/components/editorRedux.js]: Can we avoid having this as a global? First thought (without looking at the code to much) is to have an initial action that populates the store with it.

File: apps/src/sites/studio/pages/components/LevelTokenDetails.jsx [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+import _ from 'lodash';
+
+const levelKeyOptions = _.map(levelKeyList, (label, value) => ({label, value: +value}));
```
Comment (Bjvanminnen) [inline apps/src/sites/studio/pages/components/LevelTokenDetails.jsx]: Feels strange to have this so far away from the one place it's used. I imagine that the argument for not doing this directly in the render is that we avoid repeated work, which might make sense.

File: apps/src/sites/studio/pages/components/LevelTokenDetails.jsx [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+        />
+        {this.containsLegacyLevel() &&
+        <div>
+          <span style={Object.assign({display: 'inline-block'}, styles.levelFieldLabel)}>Skin</span>
+          <input defaultValue={this.props.level.skin} type="text" style={styles.textInput} />
+          <div style={{float: 'right'}}>
+            <span style={Object.assign({display: 'inline-block'}, styles.levelFieldLabel)}>Video key</span>
+            <input defaultValue={this.props.level.videoKey} type="text" style={styles.textInput} />
+          </div>
+        </div>
```
Comment (Bjvanminnen) [inline apps/src/sites/studio/pages/components/LevelTokenDetails.jsx]: I would indent the continuation (i.e. all the jsx) an extra two spaces.

File: apps/src/sites/studio/pages/components/OrderControls.jsx [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+          className="fa fa-trash"
+        />
+      </div>
+    );
+  }
+});
+
+export default connect(state => ({}), dispatch => ({
+  move(type, position, direction) {
+    dispatch({type: `MOVE_${type}`, position, direction});
```
Comment (Bjvanminnen) [inline apps/src/sites/studio/pages/components/OrderControls.jsx]: In a world where we have action creators for each of our actions, this will presumably look something more like

```
if (type === 'GROUP') {
  dispatch(removeGroup(position));
} else if {
  //etc
```

File: apps/src/sites/studio/pages/components/ScriptEditor.jsx [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+    margin: '0 0 0 7px'
+  }
+};
+
+/**
+ * Component for editing course scripts.
+ */
+const ScriptEditor = React.createClass({
+  propTypes: {
+    scriptData: React.PropTypes.object.isRequired,
```
Comment (Bjvanminnen) [inline apps/src/sites/studio/pages/components/ScriptEditor.jsx]: I find that when passing in objects of data like this, it can be hard to follow what exactly is expected.

I wonder if it would be clearer to pass in each of the individual fields as individual props. (i.e. instead of scriptData, we have hidden, loginRequired, hideable_stages, etc). At the very least, it would probably be worth (a) describing the shape of this object and/or (b) destructing it in the beginning of your render, i.e. `const { hidden, loginRequired, hideable_stages } = this.props.stageData;`

File: apps/src/sites/studio/pages/components/StageCard.jsx [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+    stagesCount: React.PropTypes.number.isRequired,
+    stage: React.PropTypes.object.isRequired
+  },
+
+  getInitialState() {
+    return {currentPositions: []};
+  },
+
+  metrics(token) {
+    return ReactDOM.findDOMNode(this.refs[`levelToken${token}`]).getBoundingClientRect();
```
Comment (Bjvanminnen) [inline apps/src/sites/studio/pages/components/StageCard.jsx]: I think in latest versions of React we may not need ReactDOM.findDOMNode (this.refs.foo already gets us the dom node)

File: apps/src/sites/studio/pages/components/StageCard.jsx [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+const StageCard = React.createClass({
+  propTypes: {
+    reorderLevel: React.PropTypes.func.isRequired,
+    addLevel: React.PropTypes.func.isRequired,
+    stagesCount: React.PropTypes.number.isRequired,
+    stage: React.PropTypes.object.isRequired
+  },
+
+  getInitialState() {
+    return {currentPositions: []};
```
Comment (Bjvanminnen) [inline apps/src/sites/studio/pages/components/StageCard.jsx]: This might be personal preference, but I like to provide default values for any state I might set in getInitialState so that there's one place I can go to look and see what state this component uses

File: apps/src/sites/studio/pages/components/StageCard.jsx [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+    window.removeEventListener('mouseup', this.handleDragStop);
+  },
+
+  handleAddLevel() {
+    this.props.addLevel(this.props.stage.position);
+  },
+
+  handleLockableChanged() {
+    const state = this.refs.lockable.checked ? 'lockable' : 'not lockable';
+    console.log(`stage ${this.props.stage.position} marked as ${state}`);
```
Comment (Bjvanminnen) [inline apps/src/sites/studio/pages/components/StageCard.jsx]: Why are we just console.logging here?

Comment (Bjvanminnen) [reply apps/src/sites/studio/pages/components/StageCard.jsx]: Also, I suspect that the onChange callback gives us an arg from which we can figure out the target instead of using refs.
i.e.

```
handleLockableChanged(event) {
  const state = event.target.checked ? 'lockable' : ' not lockable';
}
```

File: apps/src/sites/studio/pages/components/StageCard.jsx [hunk 18]
```diff
@@ -0,0 +1,166 @@
...
+    e.preventDefault();
+  },
+
+  render() {
+    return (
+      <div style={styles.stageCard}>
+        <div style={styles.stageCardHeader}>
+          Stage {this.props.stage.position}: {this.props.stage.name}
+          <OrderControls
+            type="STAGE"
```
Comment (Bjvanminnen) [inline apps/src/sites/studio/pages/components/StageCard.jsx]: Could use an enum via. makeEnum in utils for type instead of using a string, i.e. `type={ControlsType.Stage}`

File: apps/src/sites/studio/pages/components/editorRedux.js [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+
+import _ from 'lodash';
+
+function updatePositions(node) {
+  for (var i = 0; i < node.length; i++) {
```
Comment (Bjvanminnen) [inline apps/src/sites/studio/pages/components/editorRedux.js]: could be let

File: apps/src/sites/studio/pages/components/editorRedux.js [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+    node[i].position = i + 1;
+  }
+}
+
+let newStageId = -1;
+let newLevelId = -1;
```
Comment (Bjvanminnen) [inline apps/src/sites/studio/pages/components/editorRedux.js]: Your reducers are no longer pure, which is an important part of redux. Can you some how derive these from existing state instead?

File: apps/src/sites/studio/pages/components/editorRedux.js [hunk 21]
```diff
@@ -0,0 +1,127 @@
...
+
+export default function reducer(state, action) {
+  let newState = _.cloneDeep(state);
```
Comment (Bjvanminnen) [inline apps/src/sites/studio/pages/components/editorRedux.js]: This reducer is going to clone your state and return a new state for every action it sees. In your particular case, that might be fine, but if this reducer was part of a larger store, there would be lots of actions for which this reducer isn't changing anything.

A better practice would be to have each individual case statement return a newState, and if we don't hit any of the case statements just return the unchanged `state`

File: apps/src/sites/studio/pages/components/editorRedux.js [hunk 22]
```diff
@@ -0,0 +1,127 @@
+/* global levelKeyList */
+
+import _ from 'lodash';
+
+function updatePositions(node) {
```
Comment (Bjvanminnen) [inline apps/src/sites/studio/pages/components/editorRedux.js]: Should this be `incrementPositions` instead?

File: apps/src/sites/studio/pages/components/editorRedux.js [hunk 23]
```diff
@@ -0,0 +1,127 @@
...
+  for (var i = 0; i < node.length; i++) {
+    node[i].position = i + 1;
+  }
+}
+
+let newStageId = -1;
+let newLevelId = -1;
+
+export default function reducer(state, action) {
```
Comment (Bjvanminnen) [inline apps/src/sites/studio/pages/components/editorRedux.js]: I'm surprised you don't get a warning about needing an initial state. I thought redux usually complained if calling `reducer(undefined, action)` returns undefined. I would make this `reducer(state=[], action) {`

File: apps/src/sites/studio/pages/components/editorRedux.js [hunk 24]
```diff
@@ -0,0 +1,127 @@
...
+      const temp = levels.splice(action.levelA - 1, 1);
+      levels.splice(action.levelB - 1, 0, temp[0]);
+      updatePositions(levels);
+      break;
+    }
+    case 'ADD_GROUP': {
+      newState.push({
+        id: newStageId--,
+        flex_category: prompt('Enter new group name'),
+        name: prompt('Enter new stage name'),
```
Comment (Bjvanminnen) [inline apps/src/sites/studio/pages/components/editorRedux.js]: Having prompts in our reducer seems bad. We should prompt before getting here, and have the result of the prompts passed via our action.

File: apps/src/sites/studio/pages/components/StageCard.jsx (hunk 15 above)
Comment (pcardune) [reply apps/src/sites/studio/pages/components/StageCard.jsx]: Using string based refs is considered legacy, although it is not officially deprecated. See https://facebook.github.io/react/docs/refs-and-the-dom.html#the-ref-callback-attribute for an explanation of how you are "supposed" to do it.

File: apps/src/sites/studio/pages/components/editorRedux.js [hunk 25]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+let newStageId = -1;
+let newLevelId = -1;
+
+export default function reducer(state, action) {
+  let newState = _.cloneDeep(state);
+
+  switch (action.type) {
+    case 'REORDER_LEVEL': {
```
Comment (pcardune) [inline apps/src/sites/studio/pages/components/editorRedux.js]: consider prefixing these action type strings with something indicating what domain these actions are relevant for. This makes it easier to combine multiple reducers in a larger application without worrying about action type collisions.

File: apps/src/sites/studio/pages/components/editorRedux.js (hunk 23 above)
Comment (pcardune) [reply apps/src/sites/studio/pages/components/editorRedux.js]: I bet `_.cloneDeep(undefined)` returns an object.

File: apps/src/sites/studio/pages/components/editorRedux.js [hunk 26]
```diff
@@ -0,0 +1,127 @@
+/* global levelKeyList */
+
+import _ from 'lodash';
+
+function updatePositions(node) {
+  for (var i = 0; i < node.length; i++) {
+    node[i].position = i + 1;
```
Comment (Bjvanminnen) [inline apps/src/sites/studio/pages/components/editorRedux.js]: Why do we need to have `node[i].position` if it can already be derived based on it's position in the array?

File: apps/src/sites/studio/pages/components/editorRedux.js (hunk 21 above)
Comment (Bjvanminnen) [reply apps/src/sites/studio/pages/components/editorRedux.js]: The fact that most of your actions require us to clone the entire state, and then imperatively make updates also sets off red flags. Ideally, each action should only be mutating a portion of the state.

I wonder if you can construct your state and/or actions differently so that we are mutating less

File: apps/src/sites/studio/pages/components/editorRedux.js [hunk 27]
```diff
@@ -0,0 +1,127 @@
...
+      let categories = newState.map(s => s.flex_category);
+      let start = categories.indexOf(groupName);
+      let count = categories.filter(c => c === groupName).length;
+      const swap = newState.splice(start, count);
+      categories = newState.map(s => s.flex_category);
+      const swappedGroupName = newState[action.direction === 'up' ? index - 1 : index].flex_category;
+      start = categories.indexOf(swappedGroupName);
+      count = categories.filter(c => c === swappedGroupName).length;
+      newState.splice(action.direction === 'up' ? start : start + count, 0, ...swap);
+      console.log(newState.map(s => s.flex_category));
```
Comment (Bjvanminnen) [inline apps/src/sites/studio/pages/components/editorRedux.js]: console.log meant to be here?

File: apps/src/sites/studio/pages/components/editorRedux.js [hunk 28]
```diff
@@ -0,0 +1,127 @@
...
+      const groupName = newState[index].flex_category;
+      let categories = newState.map(s => s.flex_category);
+      let start = categories.indexOf(groupName);
+      let count = categories.filter(c => c === groupName).length;
+      const swap = newState.splice(start, count);
+      categories = newState.map(s => s.flex_category);
+      const swappedGroupName = newState[action.direction === 'up' ? index - 1 : index].flex_category;
+      start = categories.indexOf(swappedGroupName);
+      count = categories.filter(c => c === swappedGroupName).length;
+      newState.splice(action.direction === 'up' ? start : start + count, 0, ...swap);
```
Comment (Bjvanminnen) [inline apps/src/sites/studio/pages/components/editorRedux.js]: This is all a big block of dense code. Might be worth some comments explaining what's happening.

File: apps/src/sites/studio/pages/components/editorRedux.js (hunk 21 above)
Comment (pcardune) [reply apps/src/sites/studio/pages/components/editorRedux.js]: I would go one step further and suggest using `immutable-js` data structures for your state. It makes it a lot easier to efficiently create new state objects with slightly different data.

File: dashboard/app/views/scripts/_form.html.haml [hunk 29]
```diff
@@ -1,42 +1,27 @@
...
+  - if @script.nil? || @script.new_record?
+    .field
+      = f.label :name
+      = f.text_field :name
+  - else
+    = f.hidden_field :name
+  :javascript
+    const scriptData = #{(@script ? @script.summarize : {}).to_json};
+    const i18nData = #{(@script ? @script.summarize_i18n : {}).to_json};
+    const levelKeyList = #{Level.key_list.to_json};
```
Comment (Bjvanminnen) [inline dashboard/app/views/scripts/_form.html.haml]: We might be able to do something like we did here https://github.com/code-dot-org/code-dot-org/pull/11322 and avoid sticking these on window.

File: apps/src/sites/studio/pages/components/StageCard.jsx (hunk 15 above)
Comment (Bjvanminnen) [reply apps/src/sites/studio/pages/components/StageCard.jsx]: FWIW, we use these in a lot of places atm. I personally have yet to see a compelling argument as to why using a callback is better. In almost all cases, it seems much more awkward to me.

**Comment by Bjvanminnen** [comment]
Had a lot of small things, so wanted to summarize a couple of the more important pieces of feedback here:
(1) Your reducer is not a pure function, and needs to be.
(2) Mutating your entire state tree every time in your reducer is undesirable. Rethinking the shape of your state and/or using Immutable.js might be helpful here.
(3) Would be great to follow the pattern we've used elsewhere of making a `const` for each action type, and then exporting an action creator
(4) Let's avoid passing data from dashboard to apps by just sticking it on window. Passing in via a data attribute on the script tag might be a reasonable option here.

Let me know if you have questions about any of this :)

File: apps/src/sites/studio/pages/components/StageCard.jsx (hunk 15 above)
Comment (pcardune) [reply apps/src/sites/studio/pages/components/StageCard.jsx]: @Bjvanminnen Have you read through http://stackoverflow.com/questions/37468913/why-ref-string-is-legacy? A lot of stuff there (like static analysis type checkers) do not apply to us at the moment, but the way this breaks potential VM optimizations might be interesting for us depending on the specific scenario.

Anyway, I don't have strong opinions about string vs callback refs in our code base at large, though whenever I see string concatenation involved in object property lookups, I raise my eyebrows ever so slightly.

Comment (Bjvanminnen) [reply apps/src/sites/studio/pages/components/StageCard.jsx]: I'm inclined to agree with you on your point about seeing string concatenation involved in object property lookups. Reading through that SO link now.

File: apps/src/sites/studio/pages/components/FlexGroup.jsx (hunk 3 above)
Comment (joshlory) [reply apps/src/sites/studio/pages/components/FlexGroup.jsx]: Prevents the button from submitting the form: http://stackoverflow.com/a/10836076.

File: apps/src/sites/studio/pages/components/LevelTokenDetails.jsx (hunk 9 above)
Comment (joshlory) [reply apps/src/sites/studio/pages/components/LevelTokenDetails.jsx]: I followed the pattern in https://github.com/code-dot-org/code-dot-org/blob/d344587164a1671898b827852c72e0e892918d79/apps/src/publicKeyCryptography/IntegerDropdown.jsx#L6-L8.

Is there a better way to do this?

Comment (Bjvanminnen) [reply apps/src/sites/studio/pages/components/LevelTokenDetails.jsx]: Nope. Didn't realize we had this elsewhere already. This should be handled by the following line in our webpack config file. https://github.com/code-dot-org/code-dot-org/blob/staging/apps/webpack.js#L26

File: apps/src/sites/studio/pages/components/ScriptEditor.jsx (hunk 14 above)
Comment (joshlory) [reply apps/src/sites/studio/pages/components/ScriptEditor.jsx]: You're right, I'll split this up into individual props.

File: apps/src/sites/studio/pages/components/editorRedux.js (hunk 23 above)
Comment (joshlory) [reply apps/src/sites/studio/pages/components/editorRedux.js]: Looks like `_.cloneDeep(undefined)` returns `undefined`.  I'm surprised Redux needs an initial state when I'm explicitly passing in the initial state when I create the store.

Comment (Bjvanminnen) [reply apps/src/sites/studio/pages/components/editorRedux.js]: Oh, if you're passing in an initialstate when you create it, it probably doesn't. That would explain why you don't get warnings :)

**Comment by joshlory** [comment]
@pcardune I've moved the components to `lib/script-editor`.

**Comment by joshlory** [comment]
@Bjvanminnen I believe this is ready for another pass.  I've addressed most of your feedback but I'm still working on adding Immutable.js / not cloning the entire state tree for each action.

File: apps/src/lib/script-editor/FlexGroup.jsx [hunk 30]
```diff
@@ -0,0 +1,190 @@
...
+  },
+
+  handleAddStage(position) {
+    this.props.addStage(position, prompt('Enter new stage name'));
+  },
+
+  serializeStages(stages) {
+    let s = [];
+    stages.forEach(stage => {
+      let t = `stage '${stage.name}'`;
```
Comment (Bjvanminnen) [inline apps/src/lib/script-editor/FlexGroup.jsx:67]: could we have more descriptive variable names than s and t? (Here and in serializeLevel)

File: apps/src/lib/script-editor/FlexGroup.jsx [hunk 31]
```diff
@@ -0,0 +1,190 @@
...
+    }
+    let l = `${this.normalizeLevelKind(level.kind)} '${key.replace(/'/, "\\'")}'`;
+    if (active === false) {
+      l += ', active: false'
+    }
+    s.push(l);
+    return s;
+  },
+
+  normalizeLevelKind(kind) {
```
Comment (Bjvanminnen) [inline apps/src/lib/script-editor/FlexGroup.jsx:126]: Could be worth a comment describing why were doing this.

File: apps/src/lib/script-editor/FlexGroup.jsx [hunk 32]
```diff
@@ -0,0 +1,190 @@
...
+
+  render() {
+    const groups = _.groupBy(this.props.stages, stage => (stage.flex_category || 'Default'));
+    let count = 0;
+    let afterStage = 1;
+
+    return (
+      <div>
+        {_.map(groups, (stages, group) => {
+          return (
```
Comment (Bjvanminnen) [inline apps/src/lib/script-editor/FlexGroup.jsx]: nit, I think you could reduce indentation if you wanted by getting rid of the return statement and doing:

```
{_.map(groups, (stages, group) => (
  <div key={group}>
    ...
  </div>
)}
```

File: apps/src/lib/script-editor/LevelTokenDetails.jsx [hunk 33]
```diff
@@ -0,0 +1,226 @@
...
+          options={this.levelKindOptions}
+          onChange={this.handleLevelTypeSelected}
+          clearable={false}
+          arrowRenderer={ArrowRenderer}
+          style={styles.levelTypeSelect}
+        />
+        {this.containsLegacyLevel() &&
+          <div>
+            <span
+              style={Object.assign(
```
Comment (Bjvanminnen) [inline apps/src/lib/script-editor/LevelTokenDetails.jsx]: Could use object destructuring instead of Object.assign (here and a few other places)
`style={{display: 'inline-block'}, ...styles.levelFieldLabel}}`

Comment (Bjvanminnen) [reply apps/src/lib/script-editor/LevelTokenDetails.jsx]: Actually, you repeat that same block a few times, and none of this is dynamic config. Shouldn't display: inline-block just be part of styles.levelFieldLabel?

File: apps/src/lib/script-editor/LevelTokenDetails.jsx [hunk 34]
```diff
@@ -0,0 +1,226 @@
...
+              <hr style={styles.divider} />
+              <span style={styles.levelFieldLabel}>Active</span>
+              <input
+                type="radio"
+                onChange={this.handleActiveVariantChanged.bind(this, id)}
+                defaultChecked={id === this.props.level.activeId}
+                style={styles.checkbox}
+                name={`radio-${this.props.stagePosition}-${this.props.level.position}`}
+              />
+            </div>
```
Comment (Bjvanminnen) [inline apps/src/lib/script-editor/LevelTokenDetails.jsx]: I think we're a little bit inconsistent about this in our code (and styleguide might even say both are okay), but I think I've come around to agree with brad that it's better in cases like these to indent our JSX when part of a conditional continuation, i.e.

``` jsx
{this.props.somethingIsTrue &&
  <MyDiv/>
}
// instead of
{this.props.somethingIsTrue &&
<MyDiv/>
}
```

File: apps/src/lib/script-editor/StageCard.jsx [hunk 35]
```diff
@@ -0,0 +1,187 @@
...
+              defaultChecked={this.props.stage.lockable}
+              onChange={this.handleLockableChanged}
+              type="checkbox"
+              style={styles.checkbox}
+            />
+          </div>
+        </div>
+        {this.props.stage.levels.map(level =>
+          <LevelToken
+            ref={`levelToken${level.position}`}
```
Comment (Bjvanminnen) [inline apps/src/lib/script-editor/StageCard.jsx]: In the first iteration, I think Paul made the comment that string refs are considered legacy (which I was somewhat dismissive of). I wonder tho if this is a case where we could actually benefit from non-string refs. I think the way to do this would be to do something like:

``` js
ref={levelToken => this.metrics[level.position] = levelToken.getClientBoundingRect()}
// then elsewhere you can do
this.metrics[token]; // instead of this.metrics(token)
```

(I don't feel that strongly about this, but it might be good for us to start getting away from using string refs).

File: apps/src/lib/script-editor/editorRedux.js [hunk 36]
```diff
@@ -0,0 +1,235 @@
...
+function updatePositions(node) {
+  for (let i = 0; i < node.length; i++) {
+    node[i].position = i + 1;
+  }
+}
+
+export const NEW_LEVEL_ID = -1;
+
+export default function reducer(state, action) {
+  let newState = _.cloneDeep(state.stages);
```
Comment (Bjvanminnen) [inline apps/src/lib/script-editor/editorRedux.js:128]: Let me know if you run into any problems converting this to use Immutable.js

File: apps/src/sites/studio/pages/levelbuilder_edit_script.js [hunk 37]
```diff
@@ -1,131 +1,47 @@
...
-    margin: '0 0 0 7px'
-  }
-};
+const scriptData = scriptEditorData.script;
+const stages = scriptData.stages.filter(stage => stage.id).map(stage => ({
+  position: stage.position,
+  flex_category: stage.flex_category,
+  lockable: stage.lockable,
+  name: stage.name,
+  levels: stage.levels.filter(level => !level.uid || /_0$/.test(level.uid)).map(level => ({
```
Comment (Bjvanminnen) [inline apps/src/sites/studio/pages/levelbuilder_edit_script.js:18]: What is this filter doing (and why)? Might deserve a comment.

File: apps/src/lib/script-editor/LevelTokenDetails.jsx (hunk 34 above)
Comment (joshlory) [reply apps/src/lib/script-editor/LevelTokenDetails.jsx]: Good catch, I missed this indent!

File: apps/src/lib/script-editor/StageCard.jsx (hunk 35 above)
Comment (joshlory) [reply apps/src/lib/script-editor/StageCard.jsx]: Cool, this is much cleaner!

Comment (joshlory) [reply apps/src/lib/script-editor/StageCard.jsx]: I wonder if this is abusing `ref` to create unexpected side effects?

Comment (Bjvanminnen) [reply apps/src/lib/script-editor/StageCard.jsx]: I think this is the intention of ref (though I could be wrong). https://facebook.github.io/react/docs/refs-and-the-dom.html

Another option would be to behave more like in the link above, and just do `levelToken = this.levelTokens[position] = levelToken` and then in metrics do `this.levelTokens[position].getClientBoundingRect()`

**Comment by joshlory** [comment]
Synced with @Bjvanminnen offline.  Planning to merge this without the Immutable.js changes, to be added in a follow up PR.
